File: 100.相同的树.java
import java.util.LinkedList;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=100 lang=java
 *
 * [100] 相同的树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null && q != null) return false;
        if (p != null && q == null) return false;

        LinkedList<TreeNode> s1 = new LinkedList<>();
        LinkedList<TreeNode> s2 = new LinkedList<>();

        while (p != null|| !s1.isEmpty()) {
            if (p != null) {
                if (p.val != q.val) return false;
                s1.push(p);
                s2.push(q);
                p = p.left;
                q = q.left;
                if (p == null && q != null) return false;
                if (p != null && q == null) return false;
            } else {
                TreeNode p1 = s1.pop();
                TreeNode p2 = s2.pop();
                p = p1.right;
                q = p2.right;
                if (p == null && q != null) return false;
                if (p != null && q == null) return false;
            }
        }
        if (q != null || !s2.isEmpty()) return false;

        return true;
    }
}
// @lc code=end

File: 100.相同的树.py
#
# @lc app=leetcode.cn id=100 lang=python3
#
# [100] 相同的树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        s1, s2 = [], []
        s1.append(p)
        s2.append(q)
        while len(s1) > 0:
            n1 = s1.pop()
            n2 = s2.pop()
            if not n1 and not n2:
                continue
            if (not n1 and n2) or (n1 and not n2) or (n1.val != n2.val):
                return False
            s1.append(n1.left)
            s2.append(n2.left)
            s1.append(n1.right)
            s2.append(n2.right)

        return True
            



# @lc code=end

